File: src/Microsoft.DotNet.Interactive/KernelHost.cs
// Copyright (c) .NET Foundation and contributors. All rights reserved.
// Licensed under the MIT license. See LICENSE file in the project root for full license information.

#nullable enable

using System;
using System.Collections.Concurrent;
using System.Collections.Generic;
using System.Diagnostics;
using System.Reactive;
using System.Reactive.Concurrency;
using System.Threading;
using System.Threading.Tasks;
using Microsoft.DotNet.Interactive.Connection;
using Microsoft.DotNet.Interactive.Events;
using Microsoft.DotNet.Interactive.Formatting;
using Pocket;

namespace Microsoft.DotNet.Interactive;

public class KernelHost : IDisposable
{
    private readonly CancellationTokenSource _cancellationTokenSource = new();
    private readonly TaskCompletionSource<Unit> _disposed = new();
    private readonly CompositeKernel _kernel;
    private readonly IKernelCommandAndEventSender _defaultSender;
    private readonly IKernelCommandAndEventReceiver _receiver;
    private IDisposable? _kernelEventSubscription;
    private readonly Func<string, Task<ProxyKernel>> _defaultConnector;
    private EventLoopScheduler? _eventLoop;

    internal ConcurrentDictionary<string, KernelInvocationContext> ContextsByRootToken { get; } =
        new(StringComparer.OrdinalIgnoreCase);

    internal KernelHost(
        CompositeKernel kernel,
        IKernelCommandAndEventSender sender,
        IKernelCommandAndEventReceiver receiver,
        Uri hostUri)
    {
        Uri = hostUri ?? throw new ArgumentNullException(nameof(hostUri));
        _kernel = kernel;
        _defaultSender = sender;
        _receiver = receiver;
        _defaultConnector = async name =>
        {
            var connector = new DefaultKernelConnector(
                _defaultSender,
                _receiver);
            return await connector.CreateKernelAsync(name);
        };
        _kernel.SetHost(this);
    }

    private class DefaultKernelConnector
    {
        private readonly IKernelCommandAndEventSender _sender;
        private readonly IKernelCommandAndEventReceiver? _receiver;

        public DefaultKernelConnector(
            IKernelCommandAndEventSender sender,
            IKernelCommandAndEventReceiver receiver)
        {
            _sender = sender;
            _receiver = receiver;
        }

        public Task<ProxyKernel> CreateKernelAsync(string kernelName)
        {
            var proxy = new ProxyKernel(
                kernelName,
                _sender,
                _receiver,
                new Uri(_sender.RemoteHostUri, kernelName));

            return Task.FromResult(proxy);
        }
    }

    public async Task ConnectAsync()
    {
        _eventLoop = new EventLoopScheduler(a => new Thread(a)
        {
            Name = "KernelHost command dispatcher",
            IsBackground = true,
        });

        _kernelEventSubscription = _kernel.KernelEvents.Subscribe(e =>
        {
            if (e is ReturnValueProduced { Value: DisplayedValue })
            {
                return;
            }

            if (e is KernelInfoProduced kernelInfoProduced)
            {
                if (e.Command.DestinationUri is { } destinationUri &&
                    _kernel.ChildKernels.TryGetByUri(destinationUri, out var kernelByUri))
                {
                    kernelByUri.KernelInfo.UpdateSupportedKernelCommandsFrom(kernelInfoProduced.KernelInfo);
                }
                else if (e.Command.TargetKernelName is { } targetKernelName &&
                         _kernel.ChildKernels.TryGetByAlias(targetKernelName, out var kernelByName))
                {
                    kernelByName.KernelInfo.UpdateSupportedKernelCommandsFrom(kernelInfoProduced.KernelInfo);
                }
            }

            var _ = _defaultSender.SendAsync(e, _cancellationTokenSource.Token);
        });

        _receiver.Subscribe(commandOrEvent =>
        {
            if (commandOrEvent.IsParseError)
            {
                var _ = _defaultSender.SendAsync(commandOrEvent.Event, _cancellationTokenSource.Token);
            }
            else if (commandOrEvent.Command is { })
            {
                // this needs to be dispatched this way so that it does not block the current thread, which we see in certain bidirectional command scenarios (RequestInput sent by the SubmissionParser during magic command token interpolation) in stdio mode only (i.e. System.Console.In implementation details), and it has proven non-reproducible in tests.
                _eventLoop?.Schedule(async () =>
                {
                    try
                    {
                        await _kernel.SendAsync(commandOrEvent.Command, _cancellationTokenSource.Token);
                    }
                    catch (Exception e)
                    {
                        Logger.Log.Error(e);
                    }
                });
            }
        });

        var kernelInfos = new List<KernelInfo>();

        _kernel.VisitSubkernelsAndSelf(k =>
        {
            if (!k.KernelInfo.IsProxy)
            {
                kernelInfos.Add(k.KernelInfo);
            }
        });

        await _defaultSender.SendAsync(
            new KernelReady(kernelInfos.ToArray()),
            _cancellationTokenSource.Token);
    }

    public async Task ConnectAndWaitAsync()
    {
        await ConnectAsync();

        await _disposed.Task;
    }

    public void Dispose()
    {
        _eventLoop?.Dispose();
        _kernelEventSubscription?.Dispose();

        if (_cancellationTokenSource.Token.CanBeCanceled)
        {
            _cancellationTokenSource.Cancel();
            _cancellationTokenSource.Dispose();
        }

        _disposed.TrySetResult(Unit.Default);
    }

    public Uri Uri { get; }

    public async Task<ProxyKernel> ConnectProxyKernelAsync(
        string localName,
        Func<string, Task<ProxyKernel>> createKernelAsync,
        Uri remoteKernelUri,
        string[]? aliases = null)
    {
        if (createKernelAsync is null)
        {
            throw new ArgumentNullException(nameof(createKernelAsync));
        }

        var proxyKernel = await createKernelAsync(localName);

        proxyKernel.KernelInfo.RemoteUri = remoteKernelUri;

        if (aliases is not null)
        {
            proxyKernel.KernelInfo.NameAndAliases.UnionWith(aliases);
        }

        _kernel.Add(proxyKernel);

        return proxyKernel;
    }

    public async Task<ProxyKernel> ConnectProxyKernelOnDefaultConnectorAsync(
        string localName,
        Uri remoteKernelUri,
        string[]? aliases = null) =>
        await ConnectProxyKernelAsync(
            localName,
            _defaultConnector,
            remoteKernelUri,
            aliases);

    public static Uri CreateHostUriForCurrentProcessId() => CreateHostUriForProcessId(Process.GetCurrentProcess().Id);

    public static Uri CreateHostUriForProcessId(int processId) =>
        new($"kernel://pid-{processId}");

    public static Uri CreateHostUri(string name) => new($"kernel://{name}");
}

File: src/microsoft-dotnet-interactive/src/kernelHost.ts
// Copyright (c) .NET Foundation and contributors. All rights reserved.
// Licensed under the MIT license. See LICENSE file in the project root for full license information.

import { CompositeKernel } from './compositeKernel';
import * as contracts from './contracts';
import * as connection from './connection';
import { Kernel } from './kernel';
import { ProxyKernel } from './proxyKernel';
import { Logger } from './logger';
import { KernelScheduler } from './kernelScheduler';

export class KernelHost {
    private readonly _remoteUriToKernel = new Map<string, Kernel>();
    private readonly _uriToKernel = new Map<string, Kernel>();
    private readonly _kernelToKernelInfo = new Map<Kernel, contracts.KernelInfo>();
    private readonly _uri: string;
    private readonly _scheduler: KernelScheduler<contracts.KernelCommandEnvelope>;
    private _kernel: CompositeKernel;
    private _defaultConnector: connection.Connector;
    private readonly _connectors: connection.Connector[] = [];

    constructor(kernel: CompositeKernel, sender: connection.IKernelCommandAndEventSender, receiver: connection.IKernelCommandAndEventReceiver, hostUri: string) {
        this._kernel = kernel;
        this._uri = hostUri || "kernel://vscode";
        this._kernel.host = this;
        this._scheduler = new KernelScheduler<contracts.KernelCommandEnvelope>();

        this._defaultConnector = new connection.Connector({ sender, receiver });
        this._connectors.push(this._defaultConnector);
    }

    public get uri(): string {
        return this._uri;
    }

    public tryGetKernelByRemoteUri(remoteUri: string): Kernel | undefined {
        return this._remoteUriToKernel.get(remoteUri);
    }

    public trygetKernelByOriginUri(originUri: string): Kernel | undefined {
        return this._uriToKernel.get(originUri);
    }

    public tryGetKernelInfo(kernel: Kernel): contracts.KernelInfo | undefined {
        return this._kernelToKernelInfo.get(kernel);
    }

    public addKernelInfo(kernel: Kernel, kernelInfo: contracts.KernelInfo) {

        kernelInfo.uri = `${this._uri}/${kernel.name}`;//?
        this._kernelToKernelInfo.set(kernel, kernelInfo);
        this._uriToKernel.set(kernelInfo.uri, kernel);
    }

    public getKernel(kernelCommandEnvelope: contracts.KernelCommandEnvelope): Kernel {

        const uriToLookup = kernelCommandEnvelope.command.destinationUri ?? kernelCommandEnvelope.command.originUri;
        let kernel: Kernel | undefined = undefined;
        if (uriToLookup) {
            kernel = this._kernel.findKernelByUri(uriToLookup);
        }

        if (!kernel) {
            if (kernelCommandEnvelope.command.targetKernelName) {
                kernel = this._kernel.findKernelByName(kernelCommandEnvelope.command.targetKernelName);
            }
        }

        kernel ??= this._kernel;
        Logger.default.info(`Using Kernel ${kernel.name}`);
        return kernel;
    }

    public connectProxyKernelOnDefaultConnector(localName: string, remoteKernelUri?: string, aliases?: string[]): ProxyKernel {
        return this.connectProxyKernelOnConnector(localName, this._defaultConnector.sender, this._defaultConnector.receiver, remoteKernelUri, aliases);
    }

    public tryAddConnector(connector: { sender: connection.IKernelCommandAndEventSender, receiver: connection.IKernelCommandAndEventReceiver, remoteUris?: string[] }) {
        if (!connector.remoteUris) {
            this._connectors.push(new connection.Connector(connector));
            return true;
        } else {
            const found = connector.remoteUris!.find(uri => this._connectors.find(c => c.canReach(uri)));
            if (!found) {
                this._connectors.push(new connection.Connector(connector));
                return true;
            }
            return false;
        }
    }

    public connectProxyKernel(localName: string, remoteKernelUri: string, aliases?: string[]): ProxyKernel {
        this._connectors;//?
        const connector = this._connectors.find(c => c.canReach(remoteKernelUri));
        if (!connector) {
            throw new Error(`Cannot find connector to reach ${remoteKernelUri}`);
        }
        let kernel = new ProxyKernel(localName, connector.sender, connector.receiver);
        kernel.kernelInfo.remoteUri = remoteKernelUri;
        this._kernel.add(kernel, aliases);
        return kernel;
    }

    private connectProxyKernelOnConnector(localName: string, sender: connection.IKernelCommandAndEventSender, receiver: connection.IKernelCommandAndEventReceiver, remoteKernelUri?: string, aliases?: string[]): ProxyKernel {
        let kernel = new ProxyKernel(localName, sender, receiver);
        kernel.kernelInfo.remoteUri = remoteKernelUri;
        this._kernel.add(kernel, aliases);
        return kernel;
    }

    public tryGetConnector(remoteUri: string) {
        return this._connectors.find(c => c.canReach(remoteUri));
    }

    public connect() {
        this._kernel.subscribeToKernelEvents(e => {
            Logger.default.info(`KernelHost forwarding event: ${JSON.stringify(e)}`);
            this._defaultConnector.sender.send(e);
        });

        this._defaultConnector.receiver.subscribe({
            next: (kernelCommandOrEventEnvelope: connection.KernelCommandOrEventEnvelope) => {
                if (connection.isKernelCommandEnvelope(kernelCommandOrEventEnvelope)) {
                    Logger.default.info(`KernelHost dispacthing command: ${JSON.stringify(kernelCommandOrEventEnvelope)}`);
                    this._scheduler.runAsync(kernelCommandOrEventEnvelope, commandEnvelope => {
                        const kernel = this._kernel;
                        return kernel.send(commandEnvelope);
                    });
                }
            }
        });

        this._defaultConnector.sender.send({ eventType: contracts.KernelReadyType, event: {}, routingSlip: [this._kernel.kernelInfo.uri!] });

        this.publishKerneInfo();
    }

    public publishKerneInfo() {

        const events = this.getKernelInfoProduced();

        for (const event of events) {
            this._defaultConnector.sender.send(event);
        }
    }

    public getKernelInfoProduced(): contracts.KernelEventEnvelope[] {
        let events: contracts.KernelEventEnvelope[] = [];
        events.push({ eventType: contracts.KernelInfoProducedType, event: <contracts.KernelInfoProduced>{ kernelInfo: this._kernel.kernelInfo }, routingSlip: [this._kernel.kernelInfo.uri!] });

        for (let kernel of this._kernel.childKernels) {
            events.push({ eventType: contracts.KernelInfoProducedType, event: <contracts.KernelInfoProduced>{ kernelInfo: kernel.kernelInfo }, routingSlip: [kernel.kernelInfo.uri!] });
        }

        return events;
    }
}
